test_plans/individual-functions.asciidoc

```diff
@@ -0,0 +1,654 @@
+      [code]#marray<unsigned long int,{N}>#,
+      [code]#marray<unsigned long long int,{N}>#
+
+a@
+[source]
+----
+mgeninteger
+----
+   a@ [code]#marray<char,{N}>#, [code]#migeninteger#, [code]#mugeninteger#
+
+a@
+[source]
+----
+mgentype
+----
+   a@ [code]#mgenfloat#, [code]#mgeninteger#
+
+a@
+[source]
+----
+booln
+----
+   a@ [code]#marray<{N},bool>#
+
+a@
+[source]
+----
+genbool
+----
+   a@ [code]#bool#, [code]#booln#
+
+|====
+
+If the device has `aspect::fp64`, `double` type is tested, and 
+if the device has `aspect::fp16`, `sycl::half` type is tested.
+
+Reference templated implementations for `marray` functions will be added to `sycl-cts/util/math_reference.h`.
+
+== Tests
+
+=== 4.17.2. Function objects
+
+The function objects are already tested in CTS in `reduction` tests. The only aspect non-tested is
+their `<void>` specializations that are transparent function objects which deduce their parameter types and return type.
+Then the test will check that C++ conversion and promotion rules are satisfied for `<void>` specializations
+of the function objects.
+
+=== 4.17.5. Math functions
+
+`marray` tests will be added to every function from Table 175. _Math functions which work on SYCL host and device._
+
+=== 4.17.6. Integer functions
+
+`marray` tests will be added to every function from Table 178. _Integer functions which work on SYCL host and device,
+are available in the sycl namespace._ Additionally some problems in existing tests should be fixed and new tests added:
+
+* Reference implementation of `template <typename T, int N, typename R = typename std::make_unsigned<T>::type>
+sycl::vec<R, N> abs(sycl::vec<T, N> a)` should return `T`, not `R` following Table 178 definition.
+
+* Reference implementation of `template <typename T> T abs_diff(T a, T b)` should return `typename R = 
+typename std::make_unsigned<T>::type` instead following Table 178 definition. Implementation should be updated
+to avoid overflow and UB.
+
+* New function test for `geninteger ctz(geninteger x)` should be introduced along with reference implementation.
+
+* Reference implementations of `mad24` and `mul24` should be extended to cover full ranges for multipliers
+defined by specification.
+
+=== 4.17.7. Common functions
+
+`marray` tests will be added to every function from Table 179. _Common functions which work on SYCL host and device,
+are available in the sycl namespace._ Additionally some problems in existing tests should be fixed and new tests added:
+
+* Reference implementations of `genfloat degrees(genfloat radians)`, `genfloat radians(genfloat degrees)`,
+`genfloat smoothstep(genfloat edge0, genfloat edge1, genfloat x)`, and `genfloat sign(genfloat x)`
+with `sycl::half` should be added.
+
+* Reference implementations of `genfloat max(genfloat x, genfloat y)` and `genfloat min (genfloat x, genfloat y)`
+should be fixed to provide correct undefined return value behavior (if any input is infinite or NaN).
```

Comment:
> to provide correct undefined return value behavior

Could you clarify?

Reply:
Definition in the Table 179 states "If x or y are infinite or NaN, the return values are undefined." Current reference implementation does not satisfy this requirement.